### PR TITLE
tests: Simplifies image pulling tests 

### LINKED ISSUE #83397: tests: should be able to pull image from gcr.io should always pull from gcr.io
**Which jobs are failing**:

``sig-windows`` jobs primarely, like: https://testgrid.k8s.io/sig-windows#aks-engine-azure-windows-master

**Which test(s) are failing**:

    [k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull image from gcr.io [NodeConformance]

**Since when has it been failing**:

Since this commit merged: https://github.com/kubernetes/kubernetes/commit/2dab911ac296ae7c94e1191ce3aa37bbdfd974a8#diff-d511d08b2cdc3d78d9e75c3c131ed2ce

**Testgrid link**:

https://testgrid.k8s.io/sig-windows#aks-engine-azure-windows-master

**Reason for failure**:

The test should always pull the image from gcr.io (``gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1``), it is not configurable.

**Anything else we need to know**:

/sig testing
/sig windows

/area conformance

This issue was fixed once before: https://github.com/kubernetes/kubernetes/issues/79171

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind failing-test

/sig windows
/sig testing
/area conformance

**What this PR does / why we need it**:

Melds the ``pull image from gcr.io`` and ``pull image from docker hub`` tests into a single test that pulls the ``agnhost`` image from the configured ``e2eRegistry``.

This also removes the need to maintain and update the image ``gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1``. It should have been a manifest list that also includes future Windows releases, like Windows Server 1903. Additionally, the image has ~300 MB, meaning that with this change, it won't have to wait as much to spawn a pod.

Removes ``should not be able to pull non-existing image from gcr.io``, since the test ``should not be able to pull image from invalid registry`` test already exists, and both of them test the same effect: cannot spawn a pod with an image that does not exist.

**Which issue(s) this PR fixes**:

Fixes #83397

**Special notes for your reviewer**:

This issue was fixed once before: https://github.com/kubernetes/kubernetes/issues/79171

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/utils/image/manifest.go [hunk 1]
```diff
@@ -262,7 +263,9 @@ func initImageConfigs() map[int]Config {
 	configs[VolumeISCSIServer] = Config{e2eRegistry, "volume/iscsi", "2.0"}
 	configs[VolumeGlusterServer] = Config{e2eRegistry, "volume/gluster", "1.0"}
 	configs[VolumeRBDServer] = Config{e2eRegistry, "volume/rbd", "1.0.1"}
-	configs[WindowsNanoServer] = Config{e2eRegistry, "windows-nanoserver", "v1"}
+	// NOTE(claudiub): Do NOT change the value below to e2eRegistry, which is configurable. The image below
+	// needs to always be the one from gcr.io. (needed by: should be able to pull image from gcr.io)
+	configs[WindowsNanoServer] = Config{e2eGcRegistry, "windows-nanoserver", "v1"}
```
Comment (mkumatag) [inline test/utils/image/manifest.go]: can you just hardcode here directly if no plan to change to something else?

Comment (bclau) [reply test/utils/image/manifest.go]: sure.

Comment (bclau) [reply test/utils/image/manifest.go]: Done

**Comment by johnSchnake** [comment]
I'm confused by this test and the requirement here: if we want to hardcode a specific registry that means that the test can never pass in air-gapped environments. And this is tagged as NodeConformance.

I don't know much about `NodeConformance` and I know it is separate from `Conformance` but the naming would imply that the idea is similar: nodes should pass `NodeConformance` to be considered to be in a good state?

But how does that play with the idea of it being required to reach gcr.io? Why is that part of NodeConformance? (Do we have a definition of what that tag really means?)

The whole point of the configurable registries is that different testing situations may necessitate pulling from other locations for images. Why are we specifically testing gcr.io functionality? If we are, it seems like it shouldn't be part of `NodeConformance` and/or should be gated by some sort of `SkipUnlessInternet()`

**Comment by bclau** [comment]
> 
> 
> I'm confused by this test and the requirement here: if we want to hardcode a specific registry that means that the test can never pass in air-gapped environments. And this is tagged as NodeConformance.

The test itself checks if it can pull from gcr.io. In all Windows test runs, that ``e2eRegistry`` is configured to point at dockerhub, where we've published all the necessary images needed by the E2E tests. Obviously, without this PR, the test will try to pull from dockerhub, which is not the purpose of the test.

Nonetheless, Looking at [1], there's no mention that the node has to be air-gapped when running ``[NodeConformance]`` tests. If that is something that is truly required, then that can be done in a separate PR, which will also treat the other ``should be able to pull image from .*`` tests; it is independent of what this PR is trying to fix, the test will still require to pull the image anyways.

> 
> But how does that play with the idea of it being required to reach gcr.io? Why is that part of NodeConformance? (Do we have a definition of what that tag really means?)

In ``kubernetes/community`` there's only mention of Conformance tests. I don't think NodeConformance are mentioned, or at least from the searches I've done. I've only found [1].

> 
> The whole point of the configurable registries is that different testing situations may necessitate pulling from other locations for images. Why are we specifically testing gcr.io functionality? If we are, it seems like it shouldn't be part of `NodeConformance` and/or should be gated by some sort of `SkipUnlessInternet()`

Correct. This is not the only test in this fashion, there are a few others:

    [k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull from private registry with secret [NodeConformance]
    [k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull image from docker hub [NodeConformance]
    [k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull image from gcr.io [NodeConformance]
    [k8s.io] Container Runtime blackbox test when running a container with a new image should not be able to pull from private registry without secret [NodeConformance]
    [k8s.io] Container Runtime blackbox test when running a container with a new image should not be able to pull image from invalid registry [NodeConformance]
    [k8s.io] Container Runtime blackbox test when running a container with a new image should not be able to pull non-existing image from gcr.io [NodeConformance]

IMO, all those tests cover most possible image pulling scenarios. For the sake of coverage and completeness, the test I'm trying to fix also exists.

/test pull-kubernetes-e2e-gce-device-plugin-gpu

[1] https://kubernetes.io/docs/setup/best-practices/node-conformance/

**Comment by johnSchnake** [comment]
So all of the other examples you're listing (e.g. docker hub, invalid registry, private registry, etc) are configurable as far as I'm aware. The reason for doing this was to support the air-gapped testing use case.

I've heard use-cases where people have internet connectivity but block dockerhub or gcr.io etc. So each of those tests really need to gate themselves on `SkipUnlessConnectivityToURL(...)` or something. Otherwise people get false negatives. They'll think something is wrong with their nodes/clusters when things are operating normally.

I don't see what value those tests really add but I'm not too familiar with those inner code paths: do we really have special logic for each of these registries types (gcr vs dockerhub)? It makes sense to ensure we can do private pulling and have our error cases handled, but those dont need to be hardcoded and instead can be customizable to point to _different_ private/invalid registries if needed.

In addition, you mentioned this fixes some test failures in testgrid: why would that be? Isn't the default value for this registry the one that you want? How does the customizability of it hamper the testing?

TL;DR;
1) If the tests have no real value we should just remove them
2) If the tests have value and the specific registry (although named in the test name) doesnt matter it should be configurable
3) If the tests have value and the specific registry does matter for the codepath in question we can hardcode it but we should add a `SkipUnlessCanReachRegistry(...)` or tagged somehow (introducing new tags to the filename I think is a bit more contentious though)

Do you disagree with those 3 ideas?

File: test/utils/image/manifest.go [hunk 2]
```diff
@@ -262,7 +262,9 @@ func initImageConfigs() map[int]Config {
 	configs[VolumeISCSIServer] = Config{e2eRegistry, "volume/iscsi", "2.0"}
 	configs[VolumeGlusterServer] = Config{e2eRegistry, "volume/gluster", "1.0"}
 	configs[VolumeRBDServer] = Config{e2eRegistry, "volume/rbd", "1.0.1"}
-	configs[WindowsNanoServer] = Config{e2eRegistry, "windows-nanoserver", "v1"}
+	// NOTE(claudiub): Do NOT change the value below to e2eRegistry, which is configurable. The image below
+	// needs to always be the one from gcr.io. (needed by: should be able to pull image from gcr.io)
+	configs[WindowsNanoServer] = Config{"gcr.io/kubernetes-e2e-test-images", "windows-nanoserver", "v1"}
```
[the remark this answers is not in the document]
Comment (liggitt) [reply test/utils/image/manifest.go]: Agree. I would not expect hard-coded URLs in e2e tests that require Internet access.

**Comment by bclau** [comment]
> 
> 
> So all of the other examples you're listing (e.g. docker hub, invalid registry, private registry, etc) are configurable as far as I'm aware. The reason for doing this was to support the air-gapped testing use case.

Indeed, they are (although I don't know why the invalid registry was also made configurable - the test using an invalid image is supposed to fail to create the Pod; any invalid registry would do). The point of their configurability was so we can build the Windows images needed by the E2E tests (which are not hosted on gcr.io), and publish them in our registries (currently, they're hosted on dockerhub).

But those tests are setting the ``ImagePullPolicy`` to ``PullAlways``. In an air-gapped env, won't that fail automatically? Meaning that these tests are not meant for air-gapped envs?

> 
> I've heard use-cases where people have internet connectivity but block dockerhub or gcr.io etc. So each of those tests really need to gate themselves on `SkipUnlessConnectivityToURL(...)` or something. Otherwise people get false negatives. They'll think something is wrong with their nodes/clusters when things are operating normally.

Could we open up a separate issue for that? 

> 
> I don't see what value those tests really add but I'm not too familiar with those inner code paths: do we really have special logic for each of these registries types (gcr vs dockerhub)? It makes sense to ensure we can do private pulling and have our error cases handled, but those dont need to be hardcoded and instead can be customizable to point to _different_ private/invalid registries if needed.

I don't either. Docker client-wise, pulling dockerhub, gcr.io, mcr images is done in the same way. IMO, the test cases could be reduced to:

    should be able to pull from private registry with secret
    should be able to pull image
    should not be able to pull from private registry without secret
    should not be able to pull image from invalid registry

BTW, the tests were introduced here: https://github.com/kubernetes/kubernetes/issues/27007

> 
> In addition, you mentioned this fixes some test failures in testgrid: why would that be? Isn't the default value for this registry the one that you want? How does the customizability of it hamper the testing?

Simple. Currently, the image is:

    configs[WindowsNanoServer] = Config{e2eRegistry, "windows-nanoserver", "v1"}

By default, ``e2eRegistry`` is ``gcr.io/kubernetes-e2e-test-images``. But ``e2eRegistry`` is configurable, and we're configuring it to ``e2eteam``, so, in our case, the image pulled is ``e2eteam/windows-nanoserver:v1`` (``e2eteam`` is a dockerhub registry), but it should be ``gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1``.

Now, we can create the ``e2eteam/nanoserver:v1`` image, and make the test pass, but that would be a false positive - the test wouldn't be pulling from gcr.io as it should.

> 
> TL;DR;
> 
>     1. If the tests have no real value we should just remove them

We could reduce the number to just a few, as shown above. It's still a good idea to make sure that image pulling still works properly, especially from private registries(images could be prepulled before the tests are being run).

Personally, I would like to merge the tests:


    should be able to pull image from docker hub [NodeConformance]
    should be able to pull image from gcr.io [NodeConformance]

into just:

    should be able to pull image

which will pull the image from **any** configured registry. This would get rid of having to maintain / update the image ``gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1``, which will add a few extra benefits:

- the image has ~300 MB, which needs to be pulled entirely, which needlessly consumes time. We could simply an image like ``agnhost`` which is already used by most of the conformance tests anyways, so the pulling time would be ~0s.
- the image only works for Windows Server 1809, meaning that the gcr.io test will fail in WS 1903 and future releases. It should be a manifest list which would contain those versions, but the only people who could do that are the Googlers, which are quite hard to reach. This wouldn't be an issue anymore if we just have one generic image pulling test.

> 
>     2. If the tests have value and the specific registry (although named in the test name) doesnt matter it should be configurable

I don't quite get what you're trying to say. :(

> 
>     3. If the tests have value and the specific registry does matter for the codepath in question we can hardcode it but we should add a `SkipUnlessCanReachRegistry(...)` or tagged somehow (introducing new tags to the filename I think is a bit more contentious though)
> 

Sure, sounds fine.

**Review by timothysc — APPROVED** [review]
/approve 
/assign @johnSchnake 
^ to verify.

**Comment by johnSchnake** [comment]
```
+4 −36
```
^ is a great looking diff :)

As it stands this is fine with me; removes some tests with little->no value and removes a bunch of images from being required for testing at all. 👍
